editor/surface_upgrade_tool.cpp

```diff
@@ -32,58 +32,116 @@
+		_get_import_paths(EditorFileSystem::get_singleton()->get_filesystem(), paths);
+		for (const String &file_path : paths) {
+			String path = OS::get_singleton()->get_resource_dir() + file_path.replace_first("res://", "/");
+			print_verbose("Moving to trash: " + path);
+			Error err = OS::get_singleton()->move_to_trash(path);
+			if (err != OK) {
+				EditorNode::get_singleton()->add_io_error(TTR("Cannot remove:") + "\n" + file_path + "\n");
+			}
+		}
+
+		EditorNode::get_singleton()->restart_editor();
+	} else {
+		RS::get_singleton()->set_warn_on_surface_upgrade(true);
+	}
+}
+
+// Ensure that the warnings and popups are skipped.
+void SurfaceUpgradeTool::begin_upgrade() {
+	EditorSettings::get_singleton()->set_project_metadata("surface_upgrade_tool", "run_on_restart", false);
+	RS::get_singleton()->set_surface_upgrade_callback(nullptr);
+	RS::get_singleton()->set_warn_on_surface_upgrade(false);
+	popped_up = true;
+}
+
+void SurfaceUpgradeTool::finish_upgrade() {
+	EditorNode::get_singleton()->trigger_menu_option(EditorNode::FILE_CLOSE_ALL, true);
+
 	// Update all meshes here.
 	HashSet<String> paths;
-	PackedStringArray files_to_import;
-	_add_files(EditorFileSystem::get_singleton()->get_filesystem(), paths, files_to_import);
+	_add_files(EditorFileSystem::get_singleton()->get_filesystem(), paths);
 
 	EditorProgress ep("Re-saving all scenes and meshes", TTR("Upgrading All Meshes in Project"), paths.size());
 
-	ep.step(TTR("Re-importing meshes"), 0);
-	EditorFileSystem::get_singleton()->reimport_files(files_to_import);
-
-	uint32_t step = 1;
 	for (const String &file : paths) {
 		Ref<Resource> res = ResourceLoader::load(file);
-		ep.step(TTR("Attempting to re-save ") + file, step++);
+		ep.step(TTR("Attempting to re-save ") + file);
 		if (res.is_valid()) {
 			// Ignore things that fail to load.
 			ResourceSaver::save(res);
 		}
 	}
 }
 
-void SurfaceUpgradeTool::_show_popup() {
-	RS::get_singleton()->set_surface_upgrade_callback(nullptr);
-	bool accepted = EditorNode::immediate_confirmation_dialog(TTR("This project uses meshes with an outdated mesh format from previous Godot versions. The engine needs to update the format in order to use those meshes.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and automatically update and re-save all meshes and scenes. This update may take a few minutes. Upgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update each mesh in memory, but the update will not be saved. Choosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"), 500);
-	if (accepted) {
-		RS::get_singleton()->set_warn_on_surface_upgrade(false);
-		upgrade_all_meshes();
-	}
-}
-
 SurfaceUpgradeTool::SurfaceUpgradeTool() {
+	singleton = this;
 	RS::get_singleton()->set_surface_upgrade_callback(_show_popup);
```

Comment:
To avoid the issue reported in the comments above, I think what you can do is check `EditorFileSystem::is_importing()`. If it's true, then instead of immediately triggering the popup you delay it and keep track that you need to do it once the reimport has finished.

Reply:
Looks like it's not just importing that can get in the way, so this is not sufficient (though the idea should be valid in principle, I'm investigating further).